python/cudf/cudf/core/column/string.py

```diff
@@ -538,6 +538,11 @@ def as_numerical_column(self, dtype, **kwargs):
         elif mem_dtype.type is np.datetime64:
             kwargs.update(units=np.datetime_data(mem_dtype)[0])
             mem_dtype = np.dtype(np.int64)
+            if "format" not in kwargs:
+                fmt = pd.core.tools.datetimes._guess_datetime_format_for_array(
+                    self.data.to_host()
```

Comment:
The strings are copied to the host to call this Pandas function.
Perhaps create a feature issue to make a CUDA version of this API so the strings can stay on the GPU.

Reply:
It's definitely wasteful to copy all the strings to host - I am looking for a way to call this before the data makes it onto the device in the first place.

Reply:
I took a look at this this morning and tried to come up with a better solution. Ideally we'd be able to call this function or similar before the data makes it onto the GPU at all, and carry the format (which may be `none` if it can't figure it out) perhaps as an attribute of a `StringColumn` object which we could then pass along. The trouble with this approach is that since we can init a `StringColumn` from a fairly broad spectrum of different incoming structures, it becomes hard to implement a general solution. For instance we can read the format off of incoming numpy arrays, lists, or pandas series objects, but what if someone is trying to initialize a dataframe from a group of existing `nvstrings` objects, or something else whose data already lives on the device? In those cases we'd have to copy to host and investigate. 

The only other option I can see would be to allow the dataframe constructor to take the format as an argument. But it seems wrong to have a special branch inside the constructor to deal with this edge case.

Reply:
I think a format-guesser algorithm for device string is a reasonable request. It could be called optionally when the format is not known thereby only causing a performance hit in that case.

Reply:
If the column is mutable, or we end up concatenating two columns, we'd still probably need to process all values as a whole to determine the "best guess", so trying to guess ahead-of-time (before we copy to device) would not only be complicated, but potentially not provide much value. The ability to guess format on device would be ideal, but the implementation would need to be generic enough to work for multiple languages, or it probably won't find its way in to device code.

That being said, could we use cupy to jit a kernel from python to accomplish this? Just an idea. That approach might be impossible/overy-complicated.

In the mean-time, we should update the docstring to make a note that if the `fmt` is not specified, we will probably copy the data to the host to determine the best format. This will encourage consumers to format prior to creating the column (which is probably the best approach anyway).

Reply:
added

Reply:
This will be insanely expensive. Given we have the type as datetime64, it’s impossible for it to not have a resolution, no? I believe we have other areas of the code that show how to grab the resolution from the type and once we know the resolution it should be as simple as a dictionary lookup to get the appropriate formatting string.

Reply:
Sorry this is actually the reverse situation, I misinterpreted.

Reply:
So some thoughts:
- We have to do similar inference in parsing CSV / JSON, how is it handled there? @j-ieong @OlivierNV 
- If we use the host we can likely get away with copying a non-null value and inferencing on a single string instead of all of the strings. If they have mixed strings they’re gonna have a bad time anyway.
- If we’re using host inference against a single string we can maybe use PyArrow which will likely be more efficient than Pandas, but regardless should still be infinitely cheaper than doing all of the strings on the host.

Reply:
The single value approach crossed my mind as well, I'll roll with it. Although I am not sure what you mean about pyarrow, won't the time it takes to inspect a single data element on the host be fairly negligible regardless of how we do it?

Reply:
@kkraus14 I updated this to copy just a single non-null string to host and infer based off that.

Reply:
@brandon-b-miller the problem even with a single string is that Python uses Python string objects whereas PyArrow uses UTF8 strings. This means we do a UTF8 Device --> UTF8 Host --> Python String Host --> UTF8 Host --> UTF8 Device which can be expensive even for a single string.